### DISCUSSION TITLE
Issue with converting large BMP file

### DISCUSSION THREAD

**Discussion by RAmbika** [ImageMagick/ImageMagick#2227] [General]
We have a Large BMP image (1.1 MB) which is getting converted to Tiff. We are using the Image Magick library CORE_RL_Magick++_.dll version 7.0.8.67. The image resolution requested 300 dpi ,compression is 101,requested color is black/white. We are using the Resize function to  resize the image based on the resolution. We found that the same image conversion sometimes takes 20-30 seconds and sometimes 4-5 minites on a machine with High RAM(32 GB).On Some machines it takes 15-17 mins which lhave low RAM(4GB). Also observed that the image conversion creates temp files in the temp folder so i assume the image conversion serializes and desrializes to and from disk while conversion.Also CPU pegs upto 50 percent. Can you please tell us why this is happening and a possible solution or alterative

**Comment by snibgo** [comment]
How many pixels in your input? What is your platform?

I suppose you sometimes have too many pixels for your available memory, so IM caches the pixels on disk, which is much slower. See https://www.imagemagick.org/script/architecture.php#tera-pixel

The best solution, especially if you read the input multiple times, is first to convert it to mpc ...
```
magick in.bmp temp.mpc
```
... then then to read temp.mpc instead of the BMP file.

Another possibility is to force caching to disk. This will be slow, but not as slow as "accidentally" caching to disk:
```
magick -limit memory 1MB -limit map 1MB in.bmp out.bmp
```
Another possibility is to ensure you don't have applications running that consume memory (eg web browsers).

**Comment by RAmbika** [comment]
Hi My input image is a BMP file and resolution is:
2550 x 3507  Pixels (8.94 MPixels) (0.72) 

I am trying to use the mpc file approach and i am able to generate the mpc file which also generated the .cache file
.mpc file is 1KB and .cache file is 136 MB in size. Now i am trying to read the mpc file using the Magick:Image read function and it gives an exception. Can you please point me to the documentation where i can try to read the mpc file or use it.Thanks.

**Comment by snibgo** [comment]
Your input is quite small, if the cache takes only 136 MB. If this creates temp files in the temp folder, then I suspect other applications are hogging memory.

Read a MPC file in exactly the same way as BMP or PNG or JPEG or whatever.

Can you read the MPC from the command line? Eg:
```
magick temp.mpc -verbose info:
```

**Reply by RAmbika** [reply]
Hi I am able to read the mpc file but it is taking a longer time for processing .One observation is that a machine with higher RAM(in my case 32GB) performed well using the convert command directly rather than the APIs and when used on a machine with lower RAM took about 4 minutes. My final questions:

1. Is it better to use Convert.exe rather than using the magick++ library APIs? Will it be faster and if so is there any reason.
2. Will a machine of higher RAM give better performance for BMP files where output resolution of 300 dpi is required:
my arguments are as follows:

convert.exe c:\sample.bmp -auto-orient -resize 6483x8916 -density 300 -quality 50 -compress Group4 -type bilevel C:\\Sample.tif

Thanks

**Reply by RAmbika** [reply]
Also using 64 bit versions of Convert took about 1 min 20 seconds which is again a very considerable improvement.

**Comment by snibgo** [comment]
You use IM v7, so I suggest you use "magick" instead of "convert". (Not "magick convert".)

There is no particular reason for a pre-built magick binary being faster or slower than your own binary. Your own binary might be optimized for your CPU. On the other hand, you might not build with OpenCL and OpenMP. You might also make bugs that use more memory than necessary, or inefficient code that does more work than is necessary.

For a simple task that can be written as a magick command, there is probably no advantage in writing your own program.

If you exhaust available RAM then IM will be slow because it will use disk instead. The resolution in DPI ("-density") doesn't matter. What matters is the number of pixels.

Windows comes in two flavours for different CPU types: 32-bit (very old CPUs) and 64-bit. You should use the flavour suitable for your CPU.
